apps/back-end/src/dataset.ts

```diff
@@ -0,0 +1,263 @@
+  to: string; // destination property name
+  parser: PropParser<I, O>; // the parser function
+}
+
+/** A generic parser for property values.
+ *
+ * @param this: information about the property being parsed
+ * @param v: the value to parse
+ * @returns the parsed result.
+ */
+export type PropParser<I, O> = (this: PropParserInfo<I, O>, v: I) => O;
+
+/** An item from a dataset.
+ *
+ * Essentially this is a dictionary of arbitrary (typically primitive) unknown
+ * values, but there are a few properties which should be
+ * present in all cases.
+ *
+ * Note, multiple values represented by arrays are supported.
+ * Elements of these should be the same type.
+ *
+ * Note: if one of lat or lng properties are absent or undefined,
+ * both should be.
+ */
+export interface DatasetItem {
+  /** A unique identifier within the dataset for this item
+   *
+   * (Not to be confused with its index)
+   */
+  id: string;
+
+  /** An optional name, which may be displayed on the map as a tooltip */
+  name?: string;
+
+  /** A geographical latitude coordinate */
+  lat?: number;
+
+  /** A geographical longitude coordinate */
+  lng?: number;
+
+  /** Everything else! */
+  [key: string]: unknown;
+}
+
+/** Writes datasets to the filesystem in a form usable by the back-end */
+export class DatasetWriter {
+  /** Will be populated with the properties which are searchable */
+  readonly searchedProps: string[] = [];
+
+  /** Will be populated with the properties which are filterable */
+  readonly filteredProps: string[] = [];
+
+  /** Constructs a writer
+   *
+   * @param props - property definitions for the DatasetItems
+   * @param searchProp - the special property of item index objects into which the
+   * "searchable string" - the searchable text index - should be put.
+   */
+  constructor(
+    readonly props: PropDefs,
+    readonly searchProp: string = "searchString",
+  ) {
+    // Identify these searched and filtered properties
+    for (const propName in this.props) {
+      const propDef = this.props[propName];
+
+      if (propDef.search) this.searchedProps.push(propName);
+
+      if (propDef.isFiltered()) this.filteredProps.push(propName);
+    }
+  }
+
+  /** A JSON.stringify replacer for arrays of numbers
+   *
+   * Limits the decimals to 6 DP. This is as much resolution as a world map
+   * with building-level precesion needs.
+   */
+  limitDecimals(_: string, value: any): any {
+    const type = typeof value;
+    switch (type) {
+      case "number":
+        return Number(value.toFixed(6));
+      case "string":
+        return Number(Number(value).toFixed(6));
+      default:
+        return value;
+    }
+  }
+
+  /** A JSON.stringify replacer for objects.
+   *
+   * As above, but only alters  lat/lng properties
+   */
+  limitLatLngDecimals(key: string, value: any): any {
+    if (key === "lat" || key === "lng") {
+      if (value == null) return null; // note: nullish comparison
+      return Number(Number(value).toFixed(6));
```

Comment:
I think we decided 5 dp's, which equates to around 1m uncertainty in the position of the marker. 6 is probably fine for the size of data we have, but the performance gain will be a lot once we scale up the number of co-ops.
https://open-innovations.org/blog/2023-07-25-tips-for-optimising-geojson-files

Reply:
Ok, I don't recall offhand. Happy to change this to 5dp.